### DISCUSSION TITLE
[all] front/back 개발 테스트 하는법

### DISCUSSION THREAD

**Discussion by JuneParkCode** [3DPong/transcendence#129] [Announcements]
## integration-main
해당 브랜치는 `확인된 커밋` 이 존재하는 브랜치입니다.
여기서 `확인된 커밋` 이란, 현재 처음 init 상태를 제외하고, 테스트에서 성공적이었던 커밋만 남는다는 이야기입니다.
그러면 테스트는 어떻게 해야할까요?

### 테스트 & 커밋 방법
1. `integration-main` 브랜치 에서 브랜치를 하나 생성한다. (로컬에서 작업하세요.) 해당 브랜치를 `test-branch` 라고 부르겠습니다.
2. `test-branch` 에서 바꾼 코드를 가져옵니다.
  - 여기서 바꾼 코드를 적용하기 위한 방법은 두 가지가 있습니다.
  - **MERGE FROM WORKING BRANCH**
    - `*/develop` 브랜치에서 파생된 브랜치에서 작업을 하고 있을 것인데 해당 브랜치를 `working` 브랜치라고 부르겠습니다.
    - `working` 브랜치를 (어느정도 맞다고 생각하는 코드일겁니다. ) `test-branch` 쪽으로 머지합니다.
    - `integration-main`은 손상시키지 않고 `test-branch`에서 테스트가 가능합니다!
    - `test-branch` 에서 발생한 변경사항을 commit하고 이렇게 변경된 커밋들을cherry picking으로 다시 `working` 브랜치로 옮겨옵니다.
  - **`test-branch`에서 작업하기**
    - 테스트가 동반되며 코드작성이 필요한 경우에 많이 사용할 것입니다.
    - `test-branch` 를 `working` branch로 생각하면서 사용합니다.
    - 작업이 완료되면 `test-branch` 의 작업 커밋을  `*/develop` 에서 파생된 진짜 `working` 브랜치로 cherry picking 해옵니다.
3.  모든 작업이 완료되면 `test-branch`는 삭제합니다.  모든 작업 커밋은 `working` 에 반영되어있습니다.
4. `working` 브랜치를 `*/develop` 으로 PR 합니다.
5. PR이 완료되면 `*/develop` 에서 `integration-main`으로 PR을 하셔도 됩니다. (작동 가능한 환경 업데이트) 

### 주의
- `test` 브랜치에서는 절대 각자의 작업 디렉토리 이외의 영역을 건들면 안됩니다. 
  - 예를 들어, backend 작업 중에 frontend 코드를 변경 OR frontend 작업중에 backend 코드를 변경
  - nginx, makefile 등 개발 환경을 구성하는 공통 요소를 건드리거나...
  - 이런 일들이 발생하면 히스토리도 꼬이고, 문제가 생길 수 있습니다.

### Git Cherry-pick 사용 가이드

Git cherry-pick은 다른 브랜치에서 특정 커밋을 현재 브랜치로 가져오는 기능입니다. 이를 통해 원하는 변경 사항만 선택적으로 현재 브랜치에 적용할 수 있습니다. 다음은 Git cherry-pick에 대한 간단한 사용 가이드입니다.

1. 먼저, 원하는 커밋의 해시를 찾아야 합니다. 이를 위해 `git log` 명령어를 사용할 수 있습니다.
```
git log <source-branch>
```
여기서 `<source-branch>`는 커밋을 가져올 대상 브랜치입니다.

2. 원하는 커밋의 해시를 찾은 후, 현재 작업 중인 브랜치로 전환하세요.
```
git checkout <target-branch>
```
여기서 `<target-branch>`는 커밋을 적용할 대상 브랜치입니다.

3. 이제 cherry-pick 명령어를 사용하여 원하는 커밋을 현재 브랜치로 가져올 수 있습니다.
```
git cherry-pick <commit-hash>
```
여기서 `<commit-hash>`는 앞서 찾은 원하는 커밋의 해시입니다.

4. 만약 충돌이 발생한다면, 충돌을 해결하고 다음 명령어를 실행하세요.
```
git add <resolved-files>
git cherry-pick --continue
```
여기서 `<resolved-files>`는 충돌이 해결된 파일들입니다.

5. 충돌이 없거나 충돌을 해결한 후, 변경 사항을 원격 저장소에 푸시하세요.
```
git push origin <target-branch>
```

조금 불편할 지 모르겠지만, 지금 최선의 방법이라고 생각합니다. 각자의 브랜치가 각자의 작업을 의미하도록 신경써주세요.

**Comment by kyu-baek** [comment]
확인했습니다.
